bin/transform

```diff
@@ -245,6 +253,60 @@ def update_metadata(curated_gisaid_data: pd.DataFrame) -> pd.DataFrame:
     return curated_gisaid_data
 
 
+def warn_future_dates(gisaid_data: pd.DataFrame):
+    """
+    Sends a Slack message detailing any GISAID EPI ISL IDs in a given
+    *gisaid_data* DataFrame that have collection dates from the future (e.g. a
+    date > now).
+    """
+    def future_date(row: pd.Series):
+        """
+        Inserts a key-value pair in the within-scope *future_dates* dictionary
+        for every GISAID EPI ISL ID and future date, respectively.
+        """
+        for datetime_format in ('%Y-%m-%d', '%Y-%m', '%Y'):
+            try:
+                date = datetime.strptime(row.date, datetime_format)
+            except ValueError:
+                continue
+
+            if date > datetime.now():
+                future_dates[row.gisaid_epi_isl] = row.date
+
+            return
+
+    future_dates = {}
+    gisaid_data.apply(future_date, axis=1)
```

Comment:
I try to avoid functions that successively accumulate values in an external variable, as I think it makes the flow a bit more convoluted. In general, I think it's clearer to either a) use a for loop or b) use a functional chain of list comprehensions or maps and filters. I tend to prefer (b) as it more clearly signals that the expression is taking a list and producing a transformed list, but I think both are more clear than mixing functional (`apply`) and imperative (modifying an external variable) styles. This specific case is ~inconsequential and fine as-is, but it's a good example of the pattern so I wanted to mention it for you to think on.

Reply:
Yes, I thought this code was a bit smelly, too, but couldn't think of how else to build this dict. However, in my new approach, I discard the dict entirely.